plugins/builtin/target/azure-vmss/plugin/azure.go

```diff
@@ -190,3 +246,109 @@ func azureNodeIDMap(n *api.Node) (string, error) {
+	var vmNames []string
+
+	options := &armcompute.VirtualMachineScaleSetVMsClientListOptions{}
+
+	// If the VMSS mode is uniform, expand the instance view to get the instance ID.
+	if vmssMode == orchestrationModeUniform {
+		options.Expand = ptr.Of("instanceView")
+	}
+
+	pager := t.vmssVMs.NewListPager(resourceGroup, vmScaleSet, options)
+
+	for pager.More() {
+		page, err := pager.NextPage(ctx)
+		if err != nil {
+			return nil, fmt.Errorf("failed to list VMSS instances: %w", err)
+		}
+
+		for _, vm := range page.Value {
+			if vm.Name == nil {
+				continue
+			}
+
+			// Check for PowerState/running if instanceView is available.
+			// Unable to get flexible instanceView from the VMSS instance.
+			if vm.Properties != nil && vm.Properties.InstanceView != nil && vm.Properties.InstanceView.Statuses != nil {
+				for _, s := range vm.Properties.InstanceView.Statuses {
+					if s.Code != nil && *s.Code == "PowerState/running" {
+						t.logger.Debug("found healthy instance", "name", *vm.Name, "instance_id", *vm.InstanceID)
+						vmNames = append(vmNames, *vm.Name)
+						break
+					} else {
+						t.logger.Debug("skipping instance", "name", *vm.Name, "instance_id", *vm.InstanceID, "code", *s.Code)
+					}
+				}
+			} else if vmssMode == orchestrationModeFlexible {
+				// If mode is flexible, we cannot get the instanceView.
+				t.logger.Debug("adding instance by default for flexible mode", "name", *vm.Name, "instance_id", *vm.InstanceID)
+				vmNames = append(vmNames, *vm.Name)
+			} else {
+				// Defaults to previous logic with uniform scale sets.
+				t.logger.Debug("skipping instance", "id", *vm.ID, "instance_id", *vm.InstanceID)
+			}
+
+		}
+	}
+
+	return vmNames, nil
+}
+
+func (t TargetPlugin) getFlexibleReadyRemoteIDs(ctx context.Context, resourceGroup string, vms []string) (ids []string, err error) {
+
+	remoteIDs := []string{}
+
+	var wg sync.WaitGroup
+	// Manually limiting the number of concurrent requests to avoid hitting API rate limits.
+	requests := make(chan struct{}, 5)
+	mu := sync.Mutex{}
+
+	for _, vm := range vms {
+		wg.Add(1)
+		requests <- struct{}{}
+
+		go func(vm string) {
+			defer wg.Done()
+			defer func() { <-requests }()
+
+			select {
+			case <-ctx.Done():
+				return
+			default:
+				// Continue processing
+			}
+
+			// Get the instance view for the VM in the Flexible VMSS.
+			instanceView, err := t.vm.InstanceView(ctx, resourceGroup, vm, nil)
+			if err != nil {
+				t.logger.Debug("failed to get instance view for VM", "vm_name", vm, "error", err)
+				return
+			}
+
+			// Check for status codes on the instance view, only update remoteIDs with instances that have provisioned.
+			if len(instanceView.Statuses) == 0 {
+				for _, status := range instanceView.Statuses {
+					if *status.Code != "provisioningState/succeeded" {
+						t.logger.Debug("instance view status not ready", "vm_name", vm, "status_code", *status.Code)
+						return
+					}
+				}
+			}
```

Comment:
The guard at the top of this block is backwards. It should be checking `!=0` (but also you don't need that for non-zero lists, because ranging over it already takes care of that for you).

We're also returning early if _any_ status we see is not `provisioningState/succeeded`. Can we guarantee that if the instance view has that state that it's always going to be the first one we see?